packages/lesswrong/components/posts/TruncatedAuthorsList.tsx

```diff
@@ -1,114 +1,168 @@
+ * have to play with and the size of each name, and moves as many names as
+ * possible from the "scratch" div into the actual visible div.
+ *
+ * When there's not enough space for all the names we add some text saying '+ n
+ * more' where n is the number of excluded names. When no names can be shown we
+ * display some text saying 'n authors'.
+ */
 const recalculate = (
   ref: RefObject<HTMLDivElement>,
-  afterRef: RefObject<HTMLDivElement>,
-  selector: string,
+  expandContainer: RefObject<HTMLDivElement>,
+  classes: ClassesType,
   isAnon: boolean,
-  setOverflowCount: (count: number) => void,
 ) => {
-  if (!ref.current || !afterRef.current || isAnon) {
-    setOverflowCount(0);
+  if (!ref.current || !expandContainer.current || isAnon) {
     return;
   }
-  const afterWidth = afterRef.current.getBoundingClientRect().width ?? 0;
-  const containerRight = ref.current.getBoundingClientRect().right;
-  const elems: NodeListOf<HTMLSpanElement> = ref.current.querySelectorAll(selector);
-  elems[elems.length - 1].after(afterRef.current);
-  for (let i = 0; i < elems.length; ++i) {
-    elems[i].style.opacity = "100%";
+
+  // Get the 'off-screen' scratch buffer (it's actually on-screen with opacity 0)
+  const scratch = ref.current.querySelector("." + classes.scratch);
+  if (!scratch) {
+    return;
   }
-  for (let i = 1; i < elems.length; ++i) {
-    const elemBounds = elems[i].getBoundingClientRect();
-    if (elemBounds.right + afterWidth > containerRight) {
-      setOverflowCount(elems.length - i);
-      elems[i].before(afterRef.current);
-      for ( ; i < elems.length; ++i) {
-        elems[i].style.opacity = "0%";
-      }
+
+  // Find the '+ n more' node and make sure it's at the end of the scratch node
+  let more = scratch.querySelector("." + classes.more);
+  if (!more) {
+    more = ref.current.querySelector("." + classes.more);
+    if (!more) {
       return;
     }
+    scratch.appendChild(more);
+  }
+
+  // Move all the authors into the scratch node
+  const displayedAuthors = ref.current.querySelectorAll("." + classes.item);
+  for (const author of Array.from(displayedAuthors).reverse()) {
+    scratch.insertBefore(author, scratch.firstChild);
+  }
+
+  // Find how much space we have and what needs to fit there
+  const maxWidth = getMaxExpandableWidth(expandContainer.current);
+  const authors = Array.from(scratch.querySelectorAll("." + classes.item));
+  const bounds = authors.map((author) => author.getBoundingClientRect());
+
+  // Calculate how may authors we can fit
+  let width = bounds[0].width;
+  let moreCount = 0;
+  let i = 1;
+  for ( ; i < bounds.length; ++i) {
+    const newWidth = width + bounds[i].width + TRUNCATION_PADDING;
+    if (newWidth > maxWidth) {
+      const moreWidth = more?.getBoundingClientRect().width;
+      if (width + moreWidth > maxWidth) {
+        i--;
+      }
+      moreCount = bounds.length - i;
+      break;
+    }
+    width = newWidth;
```

Comment:
This seems sound, though I will say I found it hard for my brain to parse. No need to action that.